### ISSUE TITLE
list of active nodes are not populated in worker nodes

### ISSUE THREAD

**Issue by shlomialfasi** [prestodb/presto#12017]
in commit 90dfc04acd3efadc18136472051ba828b63ae7a6 - Simplify and cleanup DiscoveryNodeManager
the method refreshIfNecessary was removed from getAllNodes()/getActiveConnectorNodes()/getCoordinators which mean nodes that wasn't identified when the system was loaded will never be added to the node list.
on another aspect a node is assigned to a connector only after updateConnectorIds() is called in PrestoServer. updateConnectorIds() runs after DiscoveryNodeManager is initiated therefore the worker will not assign the current node to the connector list.

this problem doesn't happen in the coordinator since it has a schedule work that try to keep the active node list updated.

**Comment by dain** [comment]
Can you explain a bit more about why this is a problem for workers?

**Comment by shlomialfasi** [comment]
E.g. in case the worker needs to be aware of its own NodeId. 
in DiscoveryNodeManager where this value is populated the initialisation is done before node was assign to connector so we might end up without this value available.
another use case is that a worker node should be aware of other nodes in its cluster (in our connector we have such a use case). 
since this was the functionality before release 214 i was wondering why it was removed.

**Comment by dain** [comment]
> in DiscoveryNodeManager where this value is populated the initialization is done before node was assign to connector so we might end up without this value available.

Do you see this happen? I ask because in the constructor the record for the current node must be found, or it throws an exception and fails to start the node.  The old code was searching for the node record every time, but the information in current node is all statically known at startup:
```java
public interface Node
{
    HostAddress getHostAndPort();

    URI getHttpUri();

    String getNodeIdentifier();

    String getVersion();

    boolean isCoordinator();
}
```

> another use case is that a worker node should be aware of other nodes in its cluster (in our connector we have such a use case)

That makes sense.  The old only enabled the background refresh for the coordinator, and we likely need to do that for workers also.

> since this was the functionality before release 214 i was wondering why it was removed

This was not a conscious removal, since it was not clear that this was an expected "feature" of the code (and all tests including verification production runs did not find this issue).  Right now, I'm trying to figure out the behavior you are expecting, and the problems is change is causing.

**Comment by electrum** [comment]
@shlomialfasi Can you clarify this comment?

> E.g. in case the worker needs to be aware of its own NodeId.
in DiscoveryNodeManager where this value is populated the initialisation is done before node was assign to connector so we might end up without this value available.

The normal way a worker gets it's own node ID in a connector is via the `NodeManager.getCurrentNode()` method. This should always work. Are you saying there is a case where it does not?

**Comment by shlomialfasi** [comment]
to be more accurate a node can't find itself in the worker node list since when this list is constructed the node still not attached to the connectorId.
sorry for the confusion.

**Comment by dain** [comment]
@shlomialfasi does this fix the issue #12094?
